crates/amaru-ledger/src/rules/block.rs

```diff
@@ -47,42 +50,69 @@ pub enum InvalidBlockDetails {
         transaction_index: u32,
         violation: InvalidTransaction,
     },
-    UncategorizedError(String),
 }
 
 #[derive(Debug)]
-pub enum BlockValidation {
-    Valid,
+pub enum BlockValidation<A, E> {
+    Valid(A),
     Invalid(InvalidBlockDetails),
+    Err(E),
 }
 
-impl Try for BlockValidation {
-    type Output = ();
-    type Residual = InvalidBlockDetails;
+impl<A> BlockValidation<A, anyhow::Error> {
+    pub fn bail(msg: String) -> Self {
+        BlockValidation::Err(anyhow::Error::msg(msg))
+    }
 
-    fn from_output((): Self::Output) -> Self {
-        BlockValidation::Valid
+    pub fn anyhow<E>(err: E) -> Self
+    where
+        E: std::error::Error + Send + Sync + 'static,
+    {
+        BlockValidation::Err(anyhow::Error::new(err))
     }
 
-    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
+    pub fn context<C>(self, context: C) -> Self
+    where
+        C: std::fmt::Display + Send + Sync + 'static,
+    {
+        match self {
+            BlockValidation::Err(err) => BlockValidation::Err(err.context(context)),
+            BlockValidation::Invalid { .. } | BlockValidation::Valid { .. } => self,
+        }
+    }
+}
+
+impl<A, E> Termination for BlockValidation<A, E> {
```

Comment:
This is only useful for the tests? [Termination](https://doc.rust-lang.org/std/process/trait.Termination.html) semantic is `A trait for implementing arbitrary return types in the main function.` so feels a bit unnatural to have it implemented at the `BlockValidation` level?

Reply:
Indeed, this is used to satisfy some of the test code. 

Although, I don't think the semantic is particularly weird here. On the contrary.